# K Nearest Neighbors


In [1]:
#first import packages we might need
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

sns.set_style("whitegrid")

#### Load the data.

In [2]:
from sklearn.model_selection import train_test_split

In [6]:
## we'll start with the coffee data file
coffee_df = pd.read_csv("..\data\coffee.csv")
## check column names for the next step
coffee_df.columns

Index(['Country_of_Origin', 'Processing_Method', 'Aroma', 'Flavor',
       'Aftertaste', 'Acidity', 'Body', 'Balance', 'Uniformity', 'Clean_Cup',
       'Sweetness', 'Cupper_Points', 'altitude_low_meters',
       'altitude_high_meters', 'altitude_mean_meters'],
      dtype='object')

In [10]:
## separate the columns into X for predictors and y for the country
predictors = ['Aroma', 'Flavor',
       'Aftertaste', 'Acidity', 'Body', 'Balance', 'Uniformity', 'Clean_Cup',
       'Sweetness', 'Cupper_Points']
X = coffee_df[predictors].copy()

y = coffee_df['Country_of_Origin']





In [11]:
## perform the train test split
X_train, X_test, y_train, y_test = train_test_split(X.copy(), y.copy(),
                                                       shuffle=True,
                                                       random_state=214,
                                                       stratify=y, #be sure to stratify by the country
                                                       test_size=.2)

##### Create a pipeline for the model

In [15]:
#import packages we will need
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline
from sklearn.decomposition import PCA
from sklearn.model_selection import StratifiedKFold
from sklearn.metrics import confusion_matrix, precision_score, accuracy_score
from sklearn.neighbors import KNeighborsClassifier

In [16]:
countries = coffee_df.Country_of_Origin.value_counts()
labels = countries.index
countries.count()

19

In [46]:
kfold = StratifiedKFold(5, shuffle=True, random_state=14235)

knn_precs = np.zeros(5)

i = 0
for train_index, test_index in kfold.split(X_train, y_train.values):
    X_tt = X_train.iloc[train_index]
    X_ho = X_train.iloc[test_index]
    y_tt = y_train.iloc[train_index]
    y_ho = y_train.iloc[test_index]
    
    knn_pipe = Pipeline([('scale', StandardScaler()),
                    ('pca', PCA(10)), #this is a parameter that can be tuned   
                    ('knn', KNeighborsClassifier(n_neighbors = 7))]) #this is a parameter that can be tuned
    
    knn_pipe.fit(X_tt.values, y_tt.values)

    pred = knn_pipe.predict(X_ho.values)

    knn_precs[i] = precision_score(y_ho.values, pred, average = "macro") #the average can be changed as well
    conf_mat = confusion_matrix(y_ho.values, pred, labels=labels)


    i = i + 1

D:\Files\Anaconda\lib\site-packages\sklearn\metrics\_classification.py:1248: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
D:\Files\Anaconda\lib\site-packages\sklearn\metrics\_classification.py:1248: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
D:\Files\Anaconda\lib\site-packages\sklearn\metrics\_classification.py:1248: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
D:\Files\Anaconda\lib\site-packages\sklearn\metrics\_classification.py:1248: UndefinedMetricWarning: Precision is i

In [41]:
pd.DataFrame(conf_mat,
                 columns = [ 'Predicted '+labels[0], 'Predicted '+labels[1], 'Predicted '+labels[2], 'Predicted '+labels[3],
                            'Predicted '+labels[4], 'Predicted '+labels[5], 'Predicted '+labels[6], 'Predicted '+labels[7], 
                            'Predicted '+labels[8], 'Predicted '+labels[9], 'Predicted '+labels[10], 'Predicted '+labels[11],
                            'Predicted '+labels[12], 'Predicted '+labels[13], 'Predicted '+labels[14], 'Predicted '+labels[15],
                            'Predicted '+labels[16], 'Predicted '+labels[17], 'Predicted '+labels[18]], 
                             index = ['Actual '+labels[0], 'Actual '+labels[1], 'Actual '+labels[2], 'Actual '+labels[3],
                            'Actual '+labels[4], 'Actual '+labels[5], 'Actual '+labels[6], 'Actual '+labels[7], 
                            'Predicted '+labels[8], 'Predicted '+labels[9], 'Predicted '+labels[10], 'Predicted '+labels[11],
                            'Predicted '+labels[12], 'Predicted '+labels[13], 'Predicted '+labels[14], 'Predicted '+labels[15],
                            'Predicted '+labels[16], 'Predicted '+labels[17], 'Predicted '+labels[18]])



Predicted Mexico  Predicted Colombia  \
Actual Mexico                                         14                   4   
Actual Colombia                                        2                  16   
Actual Guatemala                                       4                   7   
Actual Brazil                                          4                   2   
Actual Taiwan                                          4                   3   
Actual United States (Hawaii)                          4                   2   
Actual Honduras                                        1                   0   
Actual Costa Rica                                      1                   3   
Predicted Ethiopia                                     0                   3   
Predicted Tanzania, United Republic Of                 1                   0   
Predicted Thailand                                     2                   0   
Predicted Nicaragua                                    1                   2   
Predicted Uganda                                       0                   1   
Predicted Kenya                                        0                   0   
Predicted El Salvador                                  0                   2   
Predicted Indonesia                                    1                   0   
Predicted China                                        1                   1   
Predicted Malawi                                       1                   0   
Predicted Peru                                         0                   0   

                                        Predicted Guatemala  Predicted Brazil  \
Actual Mexico                                             8                 5   
Actual Colombia                                           0                 6   
Actual Guatemala                                         11                 4   
Actual Brazil                                             2                 9   
Actual Taiwan                                             0                 3   
Actual United States (Hawaii)                             0                 2   
Actual Honduras                                           3                 2   
Actual Costa Rica                                         0                 1   
Predicted Ethiopia                                        1                 1   
Predicted Tanzania, United Republic Of                    1                 2   
Predicted Thailand                                        2                 1   
Predicted Nicaragua                                       0                 1   
Predicted Uganda                                          0                 2   
Predicted Kenya                                           1                 0   
Predicted El Salvador                                     0                 1   
Predicted Indonesia                                       0                 0   
Predicted China                                           1                 0   
Predicted Malawi                                          1                 0   
Predicted Peru                                            0                 0   

                                        Predicted Taiwan  \
Actual Mexico                                          0   
Actual Colombia                                        0   
Actual Guatemala                                       0   
Actual Brazil                                          0   
Actual Taiwan                                          0   
Actual United States (Hawaii)                          0   
Actual Honduras                                        0   
Actual Costa Rica                                      0   
Predicted Ethiopia                                     0   
Predicted Tanzania, United Republic Of                 0   
Predicted Thailand                                     0   
Predicted Nicaragua                                    1   
Predicted Uganda                      

In [42]:
accuracy_score(y_ho, pred)

0.29

## Summary
Above we can see the confusion matrix for the predictions from the K nearest neighbors model. We see that the model is able to correctly predict 14 Mexico samples, 16 Colombia samples, and 11 Guatemala samples. These are the countries with the highest number of entries, so it makes sense they are more likely to be correctly predicted. Unfortunately, the algorithm only obtained an accuracy of 29% with the parameters we tested (these were not optimized), so we need to make some improvements.

## Trying again with a smaller data set
We also created a data set with a stricter cutoff for the required number of entries per country (50 instead of 10 used above). We will try the model again with this smaller data set to see if we can improve the accuracy.

In [64]:
## import the new data set
coffee_df2 = pd.read_csv("..\data\coffee2.csv")
## create X and y again
## separate the columns into X for predictors and y for the country
X2 = coffee_df2[predictors].copy()

y2 = coffee_df2['Country_of_Origin']

In [66]:
## perform the train test split
X_train2, X_test2, y_train2, y_test2 = train_test_split(X2.copy(), y2.copy(),
                                                       shuffle=True,
                                                       random_state=214,
                                                       stratify=y2, #be sure to stratify by the country
                                                       test_size=.2)

In [67]:
countries2 = coffee_df2.Country_of_Origin.value_counts()
labels2 = countries2.index
countries2.count()

8

In [70]:
kfold = StratifiedKFold(5, shuffle=True, random_state=14235)

knn_precs2 = np.zeros(5)

i = 0
for train_index, test_index in kfold.split(X_train2, y_train2.values):
    X_tt2 = X_train2.iloc[train_index]
    X_ho2 = X_train2.iloc[test_index]
    y_tt2 = y_train2.iloc[train_index]
    y_ho2 = y_train2.iloc[test_index]
    
    knn_pipe2 = Pipeline([('scale', StandardScaler()),
                    ('pca', PCA(10)), #this is a parameter that can be tuned   
                    ('knn', KNeighborsClassifier(n_neighbors = 7))]) #this is a parameter that can be tuned
    
    knn_pipe2.fit(X_tt2.values, y_tt2.values)

    pred2 = knn_pipe2.predict(X_ho2.values)

    knn_precs2[i] = precision_score(y_ho2.values, pred2, average = "macro") #the average can be changed as well
    conf_mat2 = confusion_matrix(y_ho2.values, pred2, labels=labels2)


    i = i + 1

D:\Files\Anaconda\lib\site-packages\sklearn\metrics\_classification.py:1248: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [71]:
pd.DataFrame(conf_mat2,
                 columns = [ 'Predicted '+labels[0], 'Predicted '+labels[1], 'Predicted '+labels[2], 'Predicted '+labels[3],
                            'Predicted '+labels[4], 'Predicted '+labels[5], 'Predicted '+labels[6], 'Predicted '+labels[7]], 
                             index = ['Actual '+labels[0], 'Actual '+labels[1], 'Actual '+labels[2], 'Actual '+labels[3],
                            'Actual '+labels[4], 'Actual '+labels[5], 'Actual '+labels[6], 'Actual '+labels[7]])


Predicted Mexico  Predicted Colombia  \
Actual Mexico                                20                   3   
Actual Colombia                               4                  19   
Actual Guatemala                              9                   2   
Actual Brazil                                 3                   7   
Actual Taiwan                                 4                   2   
Actual United States (Hawaii)                 3                   3   
Actual Honduras                               2                   1   
Actual Costa Rica                             1                   6   

                               Predicted Guatemala  Predicted Brazil  \
Actual Mexico                                    6                 5   
Actual Colombia                                  2                 4   
Actual Guatemala                                13                 5   
Actual Brazil                                    1                 4   
Actual Taiwan                                    1                 3   
Actual United States (Hawaii)                    4                 1   
Actual Honduras                                  2                 2   
Actual Costa Rica                                0                 0   

                               Predicted Taiwan  \
Actual Mexico                                 1   
Actual Colombia                               0   
Actual Guatemala                              0   
Actual Brazil                                 1   
Actual Taiwan                                 2   
Actual United States (Hawaii)                 0   
Actual Honduras                               0   
Actual Costa Rica                             0   

                               Predicted United States (Hawaii)  \
Actual Mexico                                                 0   
Actual Colombia                                               0   
Actual Guatemala                                              0   
Actual Brazil                                                 3   
Actual Taiwan                                                 0   
Actual United States (Hawaii)                                 1   
Actual Honduras                                               0   
Actual Costa Rica                                             2   

                               Predicted Honduras  Predicted Costa Rica  
Actual Mexico                                   1                     1  
Actual Colombia                                 0                     0  
Actual Guatemala                                0                     0  
Actual Brazil                                   2                     0  
Actual Taiwan                                   0                     0  
Actual United States (Hawaii)                   0                     0  
Actual Honduras                                 1                     0  
Actual Costa Rica                               0                     0

In [72]:
accuracy_score(y_ho2, pred2)

0.3821656050955414

## Summary
We ran the pipeline again using the same parameters and here we obtained an accuracy of 38% compared to 29% obtained above. We can see that using a smaller data set with a stricter cutoff for country inclusion did increase the accuracy. However, this might be explained by the fact that Mexico, Colombia, and Guatemala, which are the countries with the largest number of samples, are now a larger proportion of the data. Looking at the confusion matrix above, we see that most countries are classifed as Mexico, Colombia, Guatemala, and Brazil. However, many of these classifications are incorrect. Unfortunately we don't seem to be able to use K Nearest Neighbors to predict the country of origin based on rating metrics such as aroma and balance.

--------------------------

This notebook was written for the Erd&#337;s Institute C&#337;de Data Science Boot Camp by Matthew Osborne, Ph. D., 2022.

Any potential redistributors must seek and receive permission from Matthew Tyler Osborne, Ph.D. prior to redistribution. Redistribution of the material contained in this repository is conditional on acknowledgement of Matthew Tyler Osborne, Ph.D.'s original authorship and sponsorship of the Erdős Institute as subject to the license (see License.md)